Upgrade Bootstrap to latest version (v3.3.x)

mnquintana opened the pull request:
~~100~~ ... 151 commits and ~~120~~ 126 changed files later...

This upgrades Bootstrap to the latest version – *but wait! There's more!* It also includes:
- A new, sexy navbar
- Tweaked color scheme
- Hundreds of style tweaks
- Refactored layout
- Updated select2
- Updated jQuery.dataTables
- Updated simple_form
- Updated jQuery.dotdotdot
- Removed jQuery.sticky (replaced with Bootstrap.affix)
- Tons of other stuff I don't remember anymore

Some notes:
- So much about the views is just really brittle and hacky (in styling, structure, and content). Fixing that stuff was out of scope.
- Our Sass has gotten messy. We need to refactor and reorganize (ie. there are variable and mixin definitions defined in random files instead of in `_variables.css.scss` and `_mixins.css.scss`)
- I removed some unused CSS, but that's a job for another PR
- ~~I commented out `rack-mini-profiler` in the Gemfile because it's too distracting during code review. Once review is done I'll uncomment it.~~ I uncommented to push a commit to force Travis to build.
- MOBILE VIEWS ARE BROKEN AS HELL. The app's responsive design is acceptable down to tablet widths, but once you get to phone widths Reservations becomes comically unusable. (also a job for another PR)
- We need to move vendor assets out of `app/assets` and into `vendor/assets` (also a job for another PR)
  - We need a better way of managing vendor assets. Because we already have a few (jQuery.dotdotdot for example) that we just downloaded, unzipped, and stuck in the folder, which is a terrible way to manage dependencies. We should switch vendor asset management over to [Bower](http://bower.io/) and/or [Rails Assets](https://rails-assets.org/)
- We need to post an updated screenshot to the site once this lands – Reservations looks a bit different now :smile: 

Yeah, I think that's it for now. Have fun :wink: 
(Fixes #245)

mnquintana commented:
Ah there we go, I knew something would have to fail. I'll fix the rubocop stuff first.

squidgetx commented:
wow this looks slickkk

a couple minor things: 

* do we want to change the navbar highlight animation? the blue glow doesn't match the new brown as well, plus it doesn't really mesh well with the new boostrap's super-flat style
* If there's an announcement or other flash message, or more than 1 or 2 items in the cart, I can't get to the checkout buttons on the cart unless I scroll all the way to the bottom of the catalog. This behavior is actually present in the old version still so idk if it's worth fixing now :P

mnquintana commented:
- Yeah I was thinking that too even though I changed the style of gradient for the nav links. I think I'm going to leave the "glow" effect on the Reservations brand component though – it's just too fun to give up. :stuck_out_tongue_closed_eyes: I did forget to change the blue color though now that I've changed the hue of brown, so I'll work on a better blue. It will still be some kind of light blue though – I think that color combination works well. But I'll probably change the hover effect for nav links to fit in better with the new flat style.
- Yeah, that is an issue but I don't think I'm gonna touch it in this PR. But it's definitely worth opening an issue for.

mnquintana commented:
Also I think I'm gonna need some help with these failing tests – are the failing tests just brittle (and I should update them) or should I actually try to make them pass? I'm having some trouble understanding what they are from the descriptions:

> 1) guest users when enabled behaves like unauthorized visiting protected route /reservations behaves like inaccessible to guests redirects to the signin page with errors

That is quite a sentence. :stuck_out_tongue:

orenyk commented:
You gotta love how RSpec blocks stack up :-). That's an integration test; if you made any changes to the CSS selectors it could easily break the spec and require that we update it.

From a quick look-over, it seems like you reworked how the flash messages are displayed? That would explain those tests failing. I'm hoping to dig into this over the weekend, I'll have more detailed feedback then. Hope that helps!

mnquintana commented:
@squidgetx New navbar hover effect now!

squidgetx commented:
damn that new navbar animation looks so good.

Other things:
* Sign in email/password fields are really long?
* Alignment on the signin page is kinda weird

orenyk commented:
Random notes as I click around:

* The navbar is the full width, not limited to the `container` max width. This makes the "Sign In" button lie outside of the nominal view, so it took me a second to find it.
* I agree w/ @squidgetx, the login form fields are way too long, we should shrink the width of that form. Also, the buttons don't line up with the form labels.
* After adding an item to the cart from the catalog, the "Add to Cart" button loses its button styling and just becomes a link. This must have something to do with the JS we're running on cart updates.
* The late fee cell in the equipment model show view (`/equipment_models/:id`) is kinda narrow given the font size used in the "Details" table (`/equipment_models/:id#details`). Since we're including the units in the cell (i.e. "$xxx.xx / day"), the "/ day" tends to run into a second line. Maybe we want to include the units in the heading (i.e. "Late Fee / day")?
* It actually looks like all of the forms are wider than they used to be (I checked out the new equipment object form, for reference). Are we missing a CSS override for `form_horizontal`?
  * As a side note, the "Back to List" link under the form is pretty easy to miss (`/equipment_models/:id/equipment_objects/new`)
    * It also takes you to the nested equipment object index (`/equipment_models/:id/equipment_objects`), not to your original location (`/equipment_models/:id`).
* Some of the links at the bottom of the ~~equipment model show view~~ nested equipment object index are no longer buttons, namely the "New [Category Name]" and "View All Equipment Items" links.
* Why is there a "Show Deactivated Items" link at the bottom of the ~~equipment model show view~~ nested equipment object index linking to the equipment object index? Did our partials get mixed up (i.e. the `_table.html.erb` partial in particular)?
* The application settings form is a bit messed up, mostly the boolean fields (checkout permissions and renewals). It's possible that they were just structurally different and we didn't notice until now.
* The report start and end date field labels are a bit small, we might want to kick the font size up a bit.
* The reservation edit form (`/reservations/:id/edit`) is pretty messed up, maybe it thinks it should be a `form_horizontal`?
* The markdown syntax guide modal is misaligned - on the reservations edit form it ends up being off the left side of the page.
* The TOS field on the user profile edit form (`/users/:id/edit`) is messed up... maybe it's a boolean field thing?
* The checkout (and checkin) receipt pages (`/reservations/checkout/:id`) are wayyy too narrow - we should definitely widen them (this was an issue before the upgrade, but while we're messing w/ views... :-P).
* The fixed position cart issue that @squidgetx mentioned is really annoying - I'll open up a new issue for that and hopefully we can get it in v5.2.0.

Overall it looks great, I really like the new flat-ness :-). I haven't looked at all through the code itself yet, I'll hopefully get to that in the next few days. Great work!

mnquintana commented:
So I agree with most of your feedback, but some of it is out of scope for this (already monstrous) PR. We should be careful not to let this turn into a master-UI-redesign PR, because that was never the intent. There happens to have been a minor UI-redesign as a result of completing the Bootstrap upgrade (and I did redesign a few things), but I'm not going to do redesign anything else in this PR.

> The navbar is the full width, not limited to the container max width. This makes the "Sign In" button lie outside of the nominal view, so it took me a second to find it.

That was an intentional design choice – the navbar was too cramped at fixed width. It'll just take some getting used to. :smile: 

> Some of the links at the bottom of the equipment model show view nested equipment object index are no longer buttons, namely the "New [Category Name]" and "View All Equipment Items" links.

Done! ([But they were never buttons.](https://github.com/YaleSTC/reservations/blob/master/app/views/equipment_objects/index.html.erb#L47-L48))

> After adding an item to the cart from the catalog, the "Add to Cart" button loses its button styling and just becomes a link. This must have something to do with the JS we're running on cart updates.

Done! (It was hard to write a comprehensive enough regex to catch all of these :confused:)

> The late fee cell in the equipment model show view (/equipment_models/:id) is kinda narrow given the font size used in the "Details" table (/equipment_models/:id#details). Since we're including the units in the cell (i.e. "$xxx.xx / day"), the "/ day" tends to run into a second line. Maybe we want to include the units in the heading (i.e. "Late Fee / day")?

Done!

> I agree w/ @squidgetx, the login form fields are way too long, we should shrink the width of that form. Also, the buttons don't line up with the form labels.
Hmm, yeah the button alignment isn't right. I'll take a look at that.

> It actually looks like all of the forms are wider than they used to be (I checked out the new equipment object form, for reference). Are we missing a CSS override for form_horizontal?

They are – it's actually a fundamental change in how Bootstrap's forms work. Form inputs for everything but inline forms are `width: 100%;` by default, and you control their width with the grid. I can make them less wide, but they will all be equal width still.

> As a side note, the "Back to List" link under the form is pretty easy to miss (/equipment_models/:id/equipment_objects/new)
>It also takes you to the nested equipment object index (/equipment_models/:id/equipment_objects), not to your original location (/equipment_models/:id).

I agree, but definitely out of scope.

> Why is there a "Show Deactivated Items" link at the bottom of the equipment model show view nested equipment object index linking to the equipment object index? Did our partials get mixed up (i.e. the _table.html.erb partial in particular)?

Huh, I'm not sure what's going on here. I'll take a look.

> The application settings form is a bit messed up, mostly the boolean fields (checkout permissions and renewals). It's possible that they were just structurally different and we didn't notice until now.
>...
> The TOS field on the user profile edit form (/users/:id/edit) is messed up... maybe it's a boolean field thing?

Yeah, it is some kind of boolean input thing, but interestingly, only in Firefox (which is why I didn't catch it oops :sheep:). I'll take a look.

> The report start and end date field labels are a bit small, we might want to kick the font size up a bit.

Eh, out of scope I think.

> The reservation edit form (/reservations/:id/edit) is pretty messed up, maybe it thinks it should be a form_horizontal?

Hahaha wow, there's always one – totally missed this view. I'm on it.

> The checkout (and checkin) receipt pages (/reservations/checkout/:id) are wayyy too narrow - we should definitely widen them (this was an issue before the upgrade, but while we're messing w/ views... :-P).

Definitely out of scope.

> The markdown syntax guide modal is misaligned - on the reservations edit form it ends up being off the left side of the page.

Yeah... this was a really annoying bug. Apparently modals don't like the new grid system, so we'd have to restructure the whole thing.

> The fixed position cart issue that @squidgetx mentioned is really annoying - I'll open up a new issue for that and hopefully we can get it in v5.2.0.

Hmmm I might be able to do something about it in here. We'll see.

orenyk commented:
Quick work! I'll review the changes locally tonight. I didn't mean for you to deal with all of my comments, btw, it was more of a brain dump as I traversed the views of the site looking for issues for the first time in a while. I'll spin out the out-of-scope comments into new issues later.

In terms of form width, I'd consider adding another column worth of padding to either side to shrink them a bit; they're just a little awkwardly wide at full size. Mobile-optimizing this whole app would be a *mess*.

In terms of the modal issues, is that something you're still working on? I think it's in-scope for this PR since it's a regression due to the upgrade.

Let me know when it's ready to be re-reviewed, but again, great work!

mnquintana commented:
Hmmm, so I'm considering replacing Bootstrap's [Affix plugin](http://getbootstrap.com/javascript/#affix) which I introduced in this PR with [jQuery.sticky](https://github.com/garand/sticky), the original library we were using. Even though Affix has a cleaner API and emits more events, it has introduced some display bugs I can't fix easily. If there are any flashes displayed, the cart becomes fixed offset from the top by the height of the flashes:

![screencap](https://cloud.githubusercontent.com/assets/823545/6013717/8838c38c-ab26-11e4-837c-003c362fbf37.gif)

It also sort of freaks out when it hits the bottom offset value (aka the footer). None of these were issues with jQuery.sticky, even though I'd rather not use a library with that little support when we could use Affix. Affix also required a lot more styling and config to work, uh, even this poorly. So I'm not sure – maybe this is just user error. :grin:

orenyk commented:
What if we made the flashes narrower so they only fit in the catalog space? Then the cart wouldn't be pushed down by the flashes (I assume it has something to do with the original position of the element?).

mnquintana commented:
I should be able to finish this this weekend.

mnquintana commented:
@orenyk @squidgetx This is ready for review again. :astonished: 

The only thing I haven't done is swap out Bootstrap.Affix for jQuery.Sticky – I'm still not sure.

squidgetx commented:
* App settings ; most of the checkboxes are fixed except for the last one (force users to accept TOS again?)
* I still think the login form fieds are too wide =/

orenyk commented:
Ditto to @squidgetx's first and last comments, but I'm actually ok with the login form width. It's not ideal but given how the view is structured it's fine (and consistent with the rest of the forms on the site). Other comments:

* The new announcements form (`/announcements/new`) is messed up, forgot to check it last time.
* The blackout edit form (`/blackouts/:id/edit`) is also messed up, but not the new blackout form (`/blackouts/new`), strangely
* The checkboxes for recurring blackouts (`/blackouts/new_recurring`) are block elements - any chance we can make them inline?
* **Sidenote** The print styling of reservations is pretty horrific (e.g. if you click the "Print this page" button) - should we fix this?
* The hover effect for the checkout and checkin procedures is kinda iffy, the pointy part looks a little pixel-y to me:
  ![selection_042](https://cloud.githubusercontent.com/assets/901808/6100599/a292e492-afe0-11e4-9ae0-c1e296329bd8.png)
* This seems to break checkout / checkin procedures
  * Steps to replicate: 1) create reservation for model with checkout procedures, 2) check out item (and check off all procedures), and 3) reservation notes indicate that procedures were not performed
  * **Sidernote** The notes for checking in an item after checking it out with missed procedures appears to be missing a leading "\n" as the header appears inside the bulleted list of missed procedures.

I'm out of time for tonight, I'll continue walking through it and then will look at the code, hopefully tomorrow.

orenyk commented:
Note: the last point is broken on `master`, I'll open up a new issue.

orenyk commented:
Some more notes:
* The contact form is broken (and not on `master`):
  ![selection_044](https://cloud.githubusercontent.com/assets/901808/6110372/44390b1e-b04f-11e4-910b-af91721350af.png)
* The import users (`/import_users/import`) and import equipment (`/import_equipment/import`) forms don't match the styling of the other forms in the app.

I think I've gone through all the views at this point, it's *finally* time to look at the code changes :-P.

orenyk commented on app/assets/javascripts/application.js:
```diff
@@ -70,41 +71,45 @@ $(document).ready(function() {
-    "aoColumnDefs": [
-          { "bSortable": false, "aTargets": [ "no_sort" ] }
-        ]
+  var wideDataTables = $('.datatable-wide').dataTable({
```
why do we assign this to a variable?

mnquintana replied:
No idea, I just added the `var` so it wouldn't be global

orenyk replied:
It's not used anywhere else, as far as I can tell, so let's get rid of the assignment.

orenyk commented:
Another comment, the padding at the top of the "Cart" part of the sidebar seems little much:
![selection_045](https://cloud.githubusercontent.com/assets/901808/6112158/2ffd3906-b05c-11e4-9be0-dbe3ee781f5b.png)

I tested out moving the flash partial to the `content` and `content_with_sidebar` partials to deal with our sticky / affix issue (so the flash only spans the width of the content `div`) and it seems acceptable to me, what do you think?

orenyk commented on app/assets/javascripts/application.js:
```diff
@@ -70,41 +71,45 @@ $(document).ready(function() {
+    $("#sidebarbottom").affix({
+      offset: {
+        top: 0,
+        bottom: function() {
```
I mean, the freaking out at the bottom that you [mentioned]() makes sense based on this line; I'm guessing without the flash offset you'd never have a situation where the bottom of the cart would be past the top of the footer.

mnquintana replied:
Nope, this happens even with a fixed integer bottom offset. Affix seems to struggle with calculating offsets – when looking at the inspector, I can see that when it freaks out it's toggling back and forth over and over between the two Affix states `.affix` and .`affix-bottom`.

orenyk replied:
Right, because the top is trying to stay below the flash div and the bottom is trying to stay above the footer, so we have conflicting constraints. What I was saying was that if we didn't have the flashes above the cart (e.g. the cart's `top` was fixed below the `navbar`), given our limits on the item list we'll never be in a situation (on a reasonably-sized screen) where the bottom of the cart would run over the footer. Does that make more sense?

orenyk commented on app/assets/javascripts/application.js:
```diff
@@ -170,13 +176,7 @@ $(document).ready(function() {
       ? ""
       : "Be aware that changing the reservation equipment item may have an effect on another reservation. If you set this reservation's equipment item to an item that has already been checked out, the reservations will be swapped.";
-    $('.form-actions').children('input').data('confirm', newMsg);
+    $('input[type="submit"]').data('confirm', newMsg);
```
This is out of scope, but shouldn't this code be in a separate file, maybe `/app/assets/javascripts/reservations.js`?

mnquintana replied:
There should actually be nothing in `application.js` but `require` directives IMO

orenyk replied:
Yea, I'd agree with that, maybe we should just have a general "Refactor Assets" issue where we deal with the vendor stuff and clear out `application.js`.

orenyk commented on app/assets/javascripts/jquery.dotdotdot-1.5.1.js:
```diff
@@ -1,569 +0,0 @@
-/*	
```
Wow, the vendor assets issue didn't sink in until I saw this... this is *terrible*!

mnquintana replied:
Yup :confused: I think we should make a separate issue for it.

orenyk commented on app/assets/javascripts/variables.js.erb:
```diff
@@ -1,2 +1,2 @@
 <% environment.context_class.instance_eval { include Rails.application.routes.url_helpers } %>
-var new_user = "<%= "#{Rails.application.config.action_controller.relative_url_root.to_s + quick_new_users_path}" + " #content" %>";
+var new_user = "<%= "#{Rails.application.config.action_controller.relative_url_root.to_s + quick_new_users_path}" + " #content *" %>";
```
What does the extra asterisk do here?

mnquintana replied:
I was having a problem with the quick new user modal, where it was loading in the `#content` div in its entirety from `layouts/_content.html.erb` inside of the modal, which was screwing up the modal display. So `#content *` is a standard jQuery / CSS selector that just selects all the elements inside of that div, instead of the div itself.

See http://api.jquery.com/load/#loading-page-fragments

orenyk replied:
Oh cool, that makes sense. Why wasn't it an issue before?

mnquintana replied:
It was, but with the old Bootstrap grid system we just never noticed.

orenyk replied:
Ha that figures.

orenyk commented on app/assets/stylesheets/_bootstrap_and_overrides.css.scss:
```diff
@@ -1,15 +1,17 @@
 }
 
-@import "bootstrap-responsive";
+// @import "bootstrap-responsive";
```
I assume we can just delete this line?

orenyk commented on app/assets/stylesheets/_bootstrap_overrides.scss:
```diff
@@ -25,14 +27,16 @@ $linkColor: lighten($right-2-color, 35%);  // #08c  Default link text color
+$navbar-default-color: lighten(desaturate($left-2-color, 40%), 40%);
+$navbar-default-link-color: lighten(desaturate($left-2-color, 40%), 40%);  // @grayLight
+$navbar-default-link-hover-color: lighten(desaturate($left-2-color, 40%), 60%);
 // $navbarLinkColorActive  // @navbarLinkColorHover
```
Is there a reason why we're keeping all these commented-out lines?

orenyk commented on app/assets/stylesheets/_global.css.scss:
```diff
@@ -28,10 +28,10 @@ label {
-  background-color: darken(desaturate($bodyBackground, 10%), 6%);
-  border-top: 1px solid darken(desaturate($bodyBackground, 10%), 15%);
-}
+// .form-actions {
```
I prefer the form buttons w/o the additional background so we can just delete these lines.

orenyk commented on app/assets/stylesheets/_global.css.scss:
```diff
@@ -42,7 +42,7 @@ label {
 // Make all view-all buttons float to the right
 // =============================================================================
-.form-actions .view_all { float: right; }
+// .form-actions .view_all { float: right; }
```
So it looks like all of the "View All" links that we do have aren't styled as buttons and don't have the class `.view_all` anyway, so I'm not sure what this was referring to. Can we just take it out?

orenyk commented on app/assets/stylesheets/catalog/_index.css.scss:
```diff
@@ -41,16 +41,18 @@
-	color: $btnWarningBackgroundHighlight;
+  float: right;
+  font-size: $font-size-base * 2;
+	color: $btn-warning-bg;
```
Is there a reason why this is indented so much?

orenyk commented on app/assets/stylesheets/equipment_models/_show.css.scss:
```diff
@@ -10,9 +9,10 @@ $border-att: 1px solid darken($bodyBackground, 20%);
+  margin-left: 0;
   margin-bottom: 3px;
   margin-top: 3px;
+  padding-left: 0;
```
was this your fix to #846?

mnquintana replied:
I think it was, I'm not sure now – it solved some rendering bug.

orenyk replied:
Ha ok, sounds good.

orenyk commented on app/assets/stylesheets/equipment_models/_show.css.scss:
```diff
@@ -113,17 +113,17 @@ $border-att: 1px solid darken($bodyBackground, 20%);
-	padding-top: 11px;
-	padding-bottom: 11px;
-	border-right: 1px solid #E5E5E5;
+	padding-top: 0;
```
indentation?

orenyk commented on app/assets/stylesheets/equipment_models/_show.css.scss:
```diff
@@ -132,14 +132,15 @@ section {
 	&:hover, &:active {
 		text-decoration: none;
-		color: $btnWarningBackground;
+		color: $btn-warning-bg;
```
Isn't this redundant now that we're setting the nominal color to `$btn-warninig-bg` (line 139)?

orenyk commented on app/assets/stylesheets/layouts/_application.css.scss:
```diff
@@ -1,130 +1,69 @@
+  //   // Subnav
+    .subnav {
+      position: static;
+  //     top: auto;
```
I assume we can remove all of these commented lines as well? Also, to confirm, the idea here is to get rid of the fixed subnav in the equipment model show view for small screens?

orenyk commented:
Found another issue - search results. First of all (not in scope for this PR), we're still getting duplicate results for equipment models (#933). Second, if you have minimal or no search results, and you click on the search field two times, the footer freaks out and jumps up on the page:

![selection_048](https://cloud.githubusercontent.com/assets/901808/6113306/70ad37c8-b064-11e4-82e8-6da5e57c200e.png)

Also, category results are kinda ugly, particularly the padding in that box.

orenyk commented:
The new user modal form is a bit messed up as well:

![selection_049](https://cloud.githubusercontent.com/assets/901808/6113483/c8bbd9f0-b065-11e4-93d9-7db008369200.png)

orenyk commented on app/helpers/reservations_helper.rb:
```diff
@@ -43,18 +43,16 @@ def manage_reservations_btn # rubocop:disable all
-                                                anchor: 'check_out_row'),
-              class: 'btn btn-inverse'
+      link_to 'Check-Out', manage_reservations_for_user_path(
+      @reservation.reserver.id,
```
Why the new indentation? I'm pretty sure the old one was more stylistically correct (aligning parameters of each method).

orenyk commented on app/views/announcements/_form.html.erb:
```diff
@@ -1,20 +1,29 @@
+    </div>
+  </div>
 
   <%#= f.label "Visible To" %>
```
We don't have this parameter in the `Announcement` model, let's just delete these lines.

orenyk commented on app/views/app_configs/_form.erb:
```diff
@@ -1,119 +1,121 @@
-  <%#= f.input :reservation_confirmation_email_active, label: 'Enable reservation confirmation?',
-              hint: 'When enabled, patrons will get an email confirmation containing the details of their reservation.' %>
-  <hr />
+    <%#= f.input :reservation_confirmation_email_active, label: 'Enable reservation confirmation?',
```
More code I think we can just delete.

orenyk commented on app/views/blackouts/_table.html.erb:
```diff
@@ -1,7 +1,7 @@
   <thead>
     <tr>
-      <!-- <th>Equipment model</th> --> <%# Setting black-outs on a per-model basis has been disabled %>
+      <%# <th>Equipment model</th> %><%# Setting black-outs on a per-model basis has been disabled %>
```
More old code

orenyk commented on app/views/blackouts/new_recurring.html.erb:
```diff
@@ -1,16 +1,11 @@
+  <%= hidden_field_tag "recurring", true %>
+  <%= f.input :days, collection: days_of_the_week_short_with_index,
+    value_method: :first, label_method: :last, as: :check_boxes,
+    item_wrapper_class: 'inline', required: true %>
```
these are still not inline (commit `6b1d84b`)

mnquintana replied:
I probably won't attempt to make these inline again. It's too difficult (related thought, I really hate simple_form :angry:)

Could totally be a separate issue though

orenyk replied:
Got it, I can't believe that this is so tricky!

mnquintana replied:
It's so difficult because simple_form's DSL is confusing and poorly documented, and frankly I'm not convinced a form builder *needs* a DSL (maybe I'd feel differently if it had better documentation). (Note: There is no better alternative – they're all like that.)

orenyk replied:
Yea, when I was trying to decipher the bootstrap initializer I had a really hard time finding documentation. One thing to note - you added the `inline` class to the checkbox `span`s but didn't actually specify that the class should have the `display: inline-block` attribute. Adding the following somewhere should fix this:

```
.checkbox.inline {
  display: inline-block;
  padding-right: 15px;
}
```

orenyk commented on app/views/cart_js/cart_dates_reload.js.erb:
```diff
@@ -5,7 +5,7 @@
 
 <% if (flash[:error].blank? && flash[:notice].blank? && flash[:alert].blank?) %>
-  $('#flashes').replaceWith('<div id="flashes"></div>');
+  $('#flashes').html('');
```
nice, much better :-)

orenyk commented on app/views/contact/new.html.erb:
```diff
@@ -1,16 +1,14 @@
-<div class="form-actions">
-  <%= f.button :submit, "Send Message" %>
-</div>
+  <%= f.button :submit, 'Send Message' %>
```
This is still raising a `NoMethodError` (commit `6b1d84b`).

mnquintana replied:
Have you restarted the server since pulling? Since it was a change in code that's only loaded on server start, you would still get the same error until you restart the server.

orenyk replied:
Whoops, good call, nice job :-).

orenyk commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -80,57 +83,58 @@
+          </li>
+        <% end %>
+      </ul>
       <% if current_user.nil? %>
```
Why wouldn't this be an `else` instead?

mnquintana replied:
Because the next conditional is in a different scope (ie. outside of the `ul`)

orenyk replied:
Right, but as I mentioned in my [other comment](https://github.com/YaleSTC/reservations/pull/1116/files#r24361008), the `ul` could be entirely within the first `if` clause (since it only applies when you're logged in) and then you can just use an `else` to display the "Sign In" `li`.

orenyk commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -80,57 +83,58 @@
-      <% # Current User Dropdown %>
-      <ul class="nav pull-right" id="logged_in">
+      <%# Current User Dropdown %>
+      <ul class="nav navbar-nav navbar-right" id="logged_in">
```
On second thought, shouldn't this be inside the `if` clause?

orenyk commented on app/views/requirements/_form.html.erb:
```diff
@@ -16,9 +16,9 @@
       <%# render 'requirement_step_fields', :f => reqstep %>
```
I feel like this whole set of lines should probably just be deleted, unless we think we'll be re-adding multi-step requirements (seems unlikely to me...).

orenyk commented on app/views/requirements/_requirement_step_fields.html.erb:
```diff
@@ -2,6 +2,6 @@
 
 <div class="requirement_steps nested-fields">
```
Yea, we've got a whole bunch of code from several versions ago; I say we clear it out (that's what git is for :-D).

orenyk commented on app/views/reservations/_cart_dates.html.erb:
```diff
@@ -7,16 +7,21 @@
-      <div class="input-append">
+      <%= hidden_field_tag 'reserver_id', cart.reserver_id %>
+      <div class="input-group">
         <%# This must all be one one line for the button to display properly, sorry it's so ugly! %>
```
looks like you resolved this comment... should we take it out?

orenyk commented on app/views/reservations/_cart_sidebar.html.erb:
```diff
@@ -6,20 +6,23 @@
+      <% end %>
     <% end %>
-    </div>
+    <%= link_to 'Empty cart'.titleize, url_for(controller: params[:controller], action: 'empty_cart'), method: :delete, class: 'btn btn-default', id: 'empty_cart_btn', remote: true %>
```
Should we float this right? Nice job switching the order of the buttons, this makes way more sense.

mnquintana replied:
I don't think we need to necessarily – it could lead to awkward spacing between the buttons.

orenyk replied:
Ok, sounds good.

orenyk commented on app/views/reservations/_form.html.erb:
```diff
@@ -1,49 +1,59 @@
+    </div>
+  </div>
 
   <% message = '' %>
```
What is this for? It doesn't seem to be used anywhere...

orenyk commented on app/views/reservations/_list_items_in_cart.html.erb:
```diff
@@ -2,12 +2,19 @@
+            <%= quantity %>
+          </td>
+          <td>
+            <div id="remove_button">
```
Much cleaner, although this will likely change via #1114.

orenyk commented on app/views/reservations/_main_page_nav_buttons.html.erb:
```diff
@@ -0,0 +1,29 @@
+<div class="btn-toolbar">
```
Where did these come from? Isn't this the old nav bar from the index view before #708? Why are we keeping this partial?

orenyk commented:
I'm not sure if this is a regression or I just wasn't paying attention before, but the "Create Reservation" page buttons are messed up:

![selection_050](https://cloud.githubusercontent.com/assets/901808/6115110/37e844ca-b071-11e4-9305-c72bf66f36ee.png)

orenyk commented on config/initializers/simple_form_bootstrap.rb:
```diff
@@ -0,0 +1,216 @@
+require 'simple_form_extensions'
```
I spent way too long trying to decipher all these options before I realized that this file was automatically generated :-P.

mnquintana replied:
Haha mostly – I had to add a bunch of stuff too.

orenyk replied:
Oh to deal with the weird checkbox alignment and whatnot?

orenyk commented:
Ok, after hours of code review I think I've gone over everything. There are/were a bunch of inline comments (as well as a few general comments for issues I found while reviewing), but I *think* that once everything's been discussed / addressed this will be good to go. Whew.

orenyk commented:
GO GO GO!

orenyk commented:
:+1:

squidgetx commented:
:shipit:

mnquintana commented:
@squidgetx @orenyk I'm running into weird stuff trying to squash this. If there were ever a PR to not squash this might be it, if only because there may actually be subcommits we want to revert since this thing is so big. Can I just merge it?

orenyk commented:
Ha, how the mighty have fallen :stuck_out_tongue:. I think it's fine in this case, @squidgetx any objections?

squidgetx commented:
nah lets go
